ParameterizedRule Migration

scottrhoyt opened the pull request:
Closes #353

scottrhoyt commented:
@jpsim I took my first stab at migrating `ParameterizedRule`s to `ConfigurableRules` using a new protocol `ViolationLevelRule` that inherits from `ConfigurableRule`. I actually really like how it has improved the readability of the rule code and reduced boilerplate, but let me know what you think.

jpsim commented on Source/SwiftLintFramework/Protocols/Rule.swift:
```diff
@@ -32,6 +32,12 @@ public protocol ConfigurableRule: Rule {
+    var error: RuleParameter<Int> { get set }
+}
+
+@available(*, deprecated=0.5.6, message="Use ConfigurableRule instead.")
```
lol, it's cute that you want to deprecate this "properly", but you should really just remove `ParameterizedRule` entirely.

We've removed things without prior deprecation notice in SwiftLint before, and I'm comfortable continuing to do that freely until SwiftLint hits 1.0.

A cursory search of GitHub seems to indicate that no one outside this project is using the protocol: https://github.com/search?l=swift&q=ParameterizedRule&ref=simplesearch&type=Code&utf8=%E2%9C%93

scottrhoyt replied:
Got it. It was more for myself to practice using that attribute and flag warnings for visibility. Planning on removing.

jpsim commented on Source/SwiftLintFramework/Protocols/Rule.swift:
```diff
@@ -59,6 +65,37 @@ public protocol CorrectableRule: Rule {
+public extension ViolationLevelRule {
+    public init?(config: AnyObject) {
+        self.init()
+        if let config = [Int].arrayOf(config) where config.count > 0 {
```
`where !config.isEmpty`

jpsim commented:
This is a fine approach, I just had a few minor comments, but otherwise you should continue down that direction.

Please note that this is the last PR blocking a 0.6.0 release :smile:

scottrhoyt commented:
@jpsim check out the last couple of changes. If nothing else is needed, ready to merge.

scottrhoyt commented:
P.S. I love seeing more deletions than additions while maintaining functionality. :+1:

jpsim commented on Source/SwiftLintFramework/Protocols/Rule.swift:
```diff
@@ -32,33 +32,46 @@ public protocol ConfigurableRule: Rule {
+        } else {
             return nil
         }
-        self.init(parameters: RuleParameter<ParameterType>.ruleParametersFromArray(array))
```
I don't think `ruleParametersFromArray` is used anymore at all, can you please remove it?

scottrhoyt replied:
Yep.

jpsim commented on Source/SwiftLintFramework/Protocols/Rule.swift:
```diff
@@ -32,33 +32,46 @@ public protocol ConfigurableRule: Rule {
-    init(parameters: [RuleParameter<ParameterType>])
-    var parameters: [RuleParameter<ParameterType>] { get }
+public protocol ViolationLevelRule: ConfigurableRule {
+    var warning: RuleParameter<Int> { get set }
```
do you need the `set`s here? If you move `self.init()` from the `ViolationLevelRule` extension to happen after setting the `warning` and `error` properties, I don't think you'll need the `set`s here.

scottrhoyt replied:
I think you do. `init?(config:)` calls `init()` first to load the default values for a `Rule` before overwriting those values with something that might be in the `config`. This is important to accommodate configurations that aren't fully specified (e.g. `variable_name_min_length: 3`). Because it is a delegating initializer, this needs to be done before you access `self`. The only way around this would be to make `init?(config:)` non-delegating. Which means we can't support both configured values and default values at the same time, it becomes an either/or. At least that is my thinking on it. Could be wrong.

jpsim replied:
you're right, let's keep as-is then.

scottrhoyt commented:
So what is odd is that as soon as I removed `ruleParametersFromArray`, I got compiler errors thrown off by `MissingDocsRule` which not only uses this method, but is a `ParameterizedRule`!!! I have no idea how this compiled on my machine or the CI server after I removed `ParameterizedRule`. It's very weird. Anyway, moved the PR back to WIP because I need to think a bit on how to best deal with `RuleParameter`s with `ParameterTypes` different than `Int`.

scottrhoyt commented:
The problem is solved for the time being by making `MissingRulesDoc` a `ConfigurableRule` straight up and keeping most of the implementation the same. Not super happy with that, but it gets us moving forward.

jpsim commented:
Great work! :clap:
